core/src/toga/hardware/camera.py

```diff
@@ -0,0 +1,155 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any
+
+from toga.constants import FlashMode
+from toga.handlers import AsyncResult
+from toga.platform import get_platform_factory
+
+if TYPE_CHECKING:
+    from toga.app import App
+
+
+class PermissionResult(AsyncResult):
+    RESULT_TYPE = "permission"
+
+
+class PhotoResult(AsyncResult):
+    RESULT_TYPE = "photo"
+
+
+class Device:
+    def __init__(self, id: str, name: str, native: Any):
+        self._id = id
+        self._name = name
+        self._native = native
+
+    @property
+    def id(self) -> str:
+        """A unique identifier for the device"""
+        return self._id
+
+    @property
+    def name(self) -> str:
+        """A human-readable name for the device"""
+        return self._name
+
+    def __eq__(self, other):
+        return self.id == other.id
+
+    def __str__(self):
+        return self._name
+
+
+class Camera:
+    def __init__(self, app: App):
+        self.factory = get_platform_factory()
+        self._app = app
+        self._impl = self.factory.Camera(self)
+
+    @property
+    def app(self) -> App:
+        """The app with which the camera is associated"""
+        return self._app
+
+    @property
+    def has_photo_permission(self) -> bool:
+        """Does the user have permission to use camera devices?
+
+        If the platform requires the user to explicitly confirm permission, and
+        the user has not given confirmation, this will prompt the user to provide
+        permission.
```

Comment:
Will it really prompt the user? That would make `has_photo_permission` basically equivalent to `request_photo_permission`, but without any way to wait until the user responds. I think it would be more useful for `has_photo_permission` to be side-effect free.

Reply:
Also, "photo" is redundant here because it's already in a "Camera" class. Suggest `has_permission` and `request_permission`, and then the same names can be used when we add other devices like GPS.

Reply:
It does prompt the user - but it's not the `has_photo_permission()` that is the trigger. On both iOS and Android, if permission hasn't been explicitly granted, the permission is asked for on the first use of a camera-enabling API. The `has_photo_permission` check in `take_photo` (a) is only there to explicitly catch the "you don't have permission" case, and (b) has an explicit through-path for the "I don't know if you have permission", which is a valid response from the underlying permission API. The public API for `camera.has_photo_permission` interprets "I don't know" as "no".

As for dropping the `photo` from `has_photo_permission` - that's to allow for video permission in future. Videos need access the camera *and* the microphone, so it's a different permission ask.

I agree that having a consistent `has_permission` API across hardware services would be desirable, but I'm not sure how we'd accomodate that when the camera has 2 possible permission sets. The only approaches I can see would be:
1. make `has_permission` a function, and have arguments for different types of permission, or
2. break the microphone out as a separate hardware device, and make the `camera.record_video` also check both `app.camera.has permission` and `app.microphone.has_permission`.

Of those, (1) doesn't strike me as much better than `has_video_permission`; (2) would have the added advantage of providing an explicit entry point for pure audio recording. Plus, we could punt implementing the `microphone` interface until we actually want to tackle audio and video playback.

Reply:
> It does prompt the user - but it's not the `has_photo_permission()` that is the trigger.

Then the second paragraph of the `has_photo_permission` documentation should be removed, because when it says "**this** will prompt the user to provide permission", what else could "this" mean except "reading the property"?

> As for dropping the `photo` from `has_photo_permission` - that's to allow for video permission in future. 

OK, makes sense.